### PR TITLE
Fix simple uses of polymorphic this in services

### LINKED ISSUE #5038: Support `this` in services
Commit: https://github.com/Microsoft/TypeScript/commit/06a250effeb19536bc7bbf7fe10c58160df02a0e disables a set of language services tests. the tests should be enabled again with the correct support.

### PR DESCRIPTION
**Pull request by sandersn**
1. Completion after `this.` was empty.
2. Quick info of methods with overloads always chose the first overload,
regardless of whether an argument whose type matched a different overload.

Both have the same cause: the type parameter introduced by polymorphic `this` is not usable, whereas the original is. In both cases, the usage is simple -- it doesn't take advantage of the capabilities of
polymorphic `this`.

Fixes #5038

### CODE REVIEW

**Comment by sandersn** [comment]
lint errors. I'll fix and push a new commit

File: src/services/services.ts [hunk 1]
```diff
@@ -4105,7 +4105,7 @@ namespace ts {
                         let useConstructSignatures = callExpression.kind === SyntaxKind.NewExpression || callExpression.expression.kind === SyntaxKind.SuperKeyword;
                         let allSignatures = useConstructSignatures ? type.getConstructSignatures() : type.getCallSignatures();
 
-                        if (!contains(allSignatures, signature.target || signature)) {
+                        if (!contains(allSignatures, signature.target) && !contains(allSignatures, signature)) {
                             // Get the first signature if there
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: "Get the first signature if there is one."

File: src/services/services.ts [hunk 2]
```diff
@@ -4105,7 +4105,7 @@ namespace ts {
                         let useConstructSignatures = callExpression.kind === SyntaxKind.NewExpression || callExpression.expression.kind === SyntaxKind.SuperKeyword;
                         let allSignatures = useConstructSignatures ? type.getConstructSignatures() : type.getCallSignatures();
 
-                        if (!contains(allSignatures, signature.target || signature)) {
+                        if (!contains(allSignatures, signature.target) && !contains(allSignatures, signature)) {
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Explain why you need to check both

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -7977,6 +7977,11 @@ namespace ts {
                 return true;
             }
             // An instance property must be accessed through an instance of the enclosing class
+            if (type.flags & TypeFlags.ThisType) {
+                // get the original type -- represented as the type constraint of the this type
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: "'this' type"

**Comment by mhegazy** [comment]
:+1:

**Comment by DanielRosenwasser** [comment]
:+1:

**Pull request merged**